### PR TITLE
Add completion callback to `Shell::AddView`

### PR DESCRIPTION
**Pull request by loic-sharma**
In the future, `FlutterEngineAddView` will be added to the embedder API to allow embedders to add views. `FlutterEngineAddView` will accept a callback that notifies the embedder once the view has been added.

This embedder API will be powered by `Shell::AddView`. This change adds a completion callback to `Shell::AddView` to prepare for the embedder API.

Design doc: https://flutter.dev/go/multi-view-embedder-apis

Part of https://github.com/flutter/flutter/issues/144806
Part of https://github.com/flutter/flutter/issues/142845

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: runtime/runtime_controller.cc [hunk 1]
```diff
@@ -136,16 +153,22 @@ bool RuntimeController::FlushRuntimeStateToIsolate() {
...
+    pending_add_view_callbacks_[view_id] = std::move(callback);
+    return;
   }
 
-  return false;
+  bool added = platform_configuration->AddView(view_id, view_metrics);
+  callback(added);
 }
 
 bool RuntimeController::RemoveView(int64_t view_id) {
```
Comment (dkwingsmt) [inline runtime/runtime_controller.cc:174]: What if a view is added and then removed before the isolate launch? I think the remove view should use the same logic as add view. And if a view is added and then removed before the isolate launch, then both the add view callback and the remove view callback should be invoked with true.

**Review by dkwingsmt — COMMENTED** [review]
See comment.

File: runtime/runtime_controller.cc (hunk 1 above)
Comment (loic-sharma) [reply runtime/runtime_controller.cc:174]: Excellent catch.

> then both the add view callback and the remove view callback should be invoked with true.

I would argue the "add callback" should be invoked with a value of `false`, no? The view was never added, the add operation was effectively canceled.

Comment (dkwingsmt) [reply runtime/runtime_controller.cc:174]: Yeah we can think this way too. And this avoids the possible tiny gap that the view controller appears "usable".

Although, then the remove view callback should also be invoked with false.

Comment (loic-sharma) [reply runtime/runtime_controller.cc:174]: Agreed. That said, this is a very weird scenario, so I could be talked into either `true` or `false` for the remove view callback 😂

File: runtime/runtime_controller.cc [hunk 2]
```diff
@@ -121,12 +121,29 @@ bool RuntimeController::FlushRuntimeStateToIsolate() {
...
+    bool added = platform_configuration->AddView(view_id, viewport_metrics);
+
+    // Callbacks will have been already invoked if the engine was restarted.
+    if (pending_add_view_callbacks_.find(view_id) !=
+        pending_add_view_callbacks_.end()) {
+      pending_add_view_callbacks_[view_id](added);
+      pending_add_view_callbacks_.erase(view_id);
+    }
+
+    if (!added) {
```
Comment (loic-sharma) [inline runtime/runtime_controller.cc:141]: Copying this conversation: https://github.com/flutter/engine/pull/51523#discussion_r1538200746

> @[dkwingsmt](https://github.com/dkwingsmt) (https://github.com/flutter/engine/pull/51523#discussion_r1538200746)
> We probably don't want to return false here but instead finish all other flushing. Imagine you added duplicate views by mistake. The app should still run except that the duplicate view controller is not usable.
> 
> @[loic-sharma](https://github.com/loic-sharma) (https://github.com/flutter/engine/pull/51523#discussion_r1538333859)
> My intuition agrees with you, but that's the pattern used by FlushRuntimeStateToIsolate today. For example if setting locales fails, the isolate's accessibility, settings, initial app lifecycle, and display info won't be flushed. I would punt changing this pattern to separate pull request if that's something we'd want to do.

File: runtime/runtime_controller.h [hunk 3]
```diff
@@ -191,6 +203,9 @@ class RuntimeController : public PlatformConfigurationClient {
   ///
   /// @param[in]  view_id  The ID of the view.
   ///
+  /// @return     If the remove view operation was forwarded to the running
+  ///             isolate.
```
Comment (loic-sharma) [inline runtime/runtime_controller.h:207]: Copying this conversation: https://github.com/flutter/engine/pull/51523#discussion_r1538203021

> @[dkwingsmt](https://github.com/dkwingsmt) (https://github.com/flutter/engine/pull/51523#discussion_r1538203021)
> This explanation is very hard to understand to me. Can it simply be "Whether the removal is successful."?
> 
> @[loic-sharma](https://github.com/loic-sharma) (https://github.com/flutter/engine/pull/51523#discussion_r1538335582)
> This is the language used by other similar methods on this class:
> 
> [engine/runtime/runtime_controller.h](https://github.com/flutter/engine/blob/b2001fa6d7738752294aea4c2fb87486cf84c174/runtime/runtime_controller.h#L204-L206)
>
> > Can it simply be "Whether the removal is successful."?
> 
> There's a subtle difference: if the Dart isolate isn't launched, this succeeds and returns false. If we want to change this behavior, I would say that's better kept to a separate pull request :)

File: runtime/runtime_controller.cc (hunk 2 above)
Comment (dkwingsmt) [reply runtime/runtime_controller.cc:141]: I don't think they're of the same case though. All current `SetXxxx` operations simply flushes the state to the platform configuration, and succeeds as long as there is a platform configuration. The `AddView`, however, also fails if the view ID is invalid, which is a user error. In short, the engine can stop initializing if the user does some wrong operation before the Dart isolate is ready, which is not something that would happen today.

File: shell/common/shell_unittests.cc [hunk 4]
```diff
@@ -4570,8 +4690,15 @@ TEST_F(ShellTest, ShellFlushesPlatformStatesByMain) {
     // The construtor for ViewportMetrics{_, width, _, _, _} (only the 2nd
     // argument matters in this test).
     platform_view->SetViewportMetrics(0, ViewportMetrics{1, 10, 1, 0, 0});
-    shell->AddView(1, ViewportMetrics{1, 30, 1, 0, 0});
+    shell->AddView(1, ViewportMetrics{1, 30, 1, 0, 0},
+                   [](bool added) { ASSERT_TRUE(added); });
     platform_view->SetViewportMetrics(0, ViewportMetrics{1, 20, 1, 0, 0});
+
+    // A view can be added and removed all before the isolate launches.
```
Comment (dkwingsmt) [inline shell/common/shell_unittests.cc:4697]: Can we move this to a separate test? The feature of adding and removing views with calling back the correct result feels enough for a standalone test.

File: runtime/runtime_controller.cc (hunk 2 above)
Comment (loic-sharma) [reply runtime/runtime_controller.cc:141]: I think this code does follow today's behavior. Today `FlushRuntimeStateToIsolate` calls `AddView` for all viewport metrics. If any `AddView` returns `false`, `FlushRuntimeStateToIsolate` immediately returns `false`:

https://github.com/flutter/engine/blob/3f332b235c8973809ef8ba9114eb6055c4800bdd/runtime/runtime_controller.cc#L124-L129

Please let me know if I'm missing something.

File: shell/common/shell_unittests.cc (hunk 4 above)
Comment (loic-sharma) [reply shell/common/shell_unittests.cc]: Sure, moved!

File: runtime/runtime_controller.cc (hunk 2 above)
Comment (dkwingsmt) [reply runtime/runtime_controller.cc:141]: No. Because `AddView` may return false for two reasons:
* `add_view_.dart_state` doesn't exist, which typically happens during the destruction of the Dart isolate, and is an **expected exception**.
* The `view_id` is illegal, which is a user **error**.

All other operations, taking `SetLocales` for example, only return false if the `platform_configuration` doesn't exist, an **expected exception**. They never return false on user errors. 

In short, in today's main, the RuntimeController will never stop initializing on user errors. With this change, it will.

Comment (loic-sharma) [reply runtime/runtime_controller.cc:141]: Ok gotcha, will fix!

Comment (loic-sharma) [reply runtime/runtime_controller.cc:141]: @dkwingsmt This is fixed now!

File: runtime/runtime_controller.cc [hunk 5]
```diff
@@ -136,25 +154,51 @@ bool RuntimeController::FlushRuntimeStateToIsolate() {
...
+  // add the view and invoke the callback when the isolate is started.
+  auto* platform_configuration = GetPlatformConfigurationIfAvailable();
+  if (!platform_configuration) {
+    FML_DCHECK(has_flushed_runtime_state_ == false);
+
+    if (pending_add_view_callbacks_.find(view_id) !=
+        pending_add_view_callbacks_.end()) {
+      FML_LOG(ERROR) << "View #" << view_id << " is already pending creation.";
+      callback(false);
+      return;
```
Comment (dkwingsmt) [inline runtime/runtime_controller.cc:170]: We should probably remove the `return`, since we want to add the other views even when one view is a user error.

File: runtime/runtime_controller.cc [hunk 6]
```diff
@@ -136,25 +154,51 @@ bool RuntimeController::FlushRuntimeStateToIsolate() {
...
+    }
 
-    return true;
+    platform_data_.viewport_metrics_for_views[view_id] = view_metrics;
+    pending_add_view_callbacks_[view_id] = std::move(callback);
+    return;
   }
```
Comment (dkwingsmt) [inline runtime/runtime_controller.cc:177]: nit: Shall we assert `pending_add_view_callbacks_` being empty here?

File: runtime/runtime_controller.h [hunk 7]
```diff
@@ -176,10 +181,17 @@ class RuntimeController : public PlatformConfigurationClient {
   ///             including the implicit view. Adding a view that already exists
   ///             triggers an assertion.
   ///
+  ///             If the isolate is not running, the view will be saved and
+  ///             flushed to the isolate when it starts.
```
Comment (dkwingsmt) [inline runtime/runtime_controller.h:185]:
```suggestion
  ///             If the isolate is not running, the view will be saved and
  ///             flushed to the isolate when it starts, and the callback
  ///             is not completed until then. If such a view is removed 
  ///             before the isolate starts, the callback is completed immediately with 
  ///             false.
```

File: runtime/runtime_controller.h (hunk 3 above)
Comment (dkwingsmt) [inline runtime/runtime_controller.h:207]:
```suggestion
  ///             isolate. If the Dart isolate is not running, then 
  ///             the pending view creation (if any) is canceled, 
  ///             and the return value is always false.
```

File: runtime/runtime_controller.h [hunk 8]
```diff
@@ -660,6 +675,11 @@ class RuntimeController : public PlatformConfigurationClient {
       std::shared_ptr<PlatformIsolateManager>(new PlatformIsolateManager());
   bool has_flushed_runtime_state_ = false;
 
+  // Callbacks when `AddView` was called before the Dart isolate is launched.
+  //
+  // These views will be added when `FlushRuntimeStateToIsolate` is called.
```
Comment (dkwingsmt) [inline runtime/runtime_controller.h:680]:
```suggestion
  // These views will be added when `FlushRuntimeStateToIsolate` is called.
  // It is no longer used once the Dart isolate starts.
```

File: runtime/runtime_controller.cc (hunk 5 above)
Comment (loic-sharma) [reply runtime/runtime_controller.cc:170]: What do you mean? This is a failure if you add the same view twice while the Dart isolate isn't launched. If before the Dart isolate is launched you add view `1` twice and then view `2`, view `2` will still be added.

**Review by dkwingsmt — APPROVED** [review]
LGTM with comment suggestions

File: runtime/runtime_controller.h (hunk 3 above)
Comment (loic-sharma) [reply runtime/runtime_controller.h]: Added!

File: runtime/runtime_controller.h (hunk 8 above)
Comment (loic-sharma) [reply runtime/runtime_controller.h:695]: Added!

File: runtime/runtime_controller.h (hunk 7 above)
Comment (loic-sharma) [reply runtime/runtime_controller.h]: I added this and tweaked the wording a bit. I also added similar content to the `AddViewCallback` declaration too.

File: runtime/runtime_controller.cc (hunk 6 above)
Comment (loic-sharma) [reply runtime/runtime_controller.cc:177]: Added an assertion for this case!
